#!/usr/bin/env Rscript
#################################################################
# PDX_bwa_mem_mapping_jobs_from_master_manif.R
# This script takes an MAster manifest generated from 6591_master_manifest_creation_RNAseq.R to create series of files with a series of command job
# submissions to submit the mapping of the fastq files to the mouse reference genome
#
#  The Project folder should have the following  structure before running this program:
#  /My_project_folder/
#   ├── metadata/manifests/My_cram_manifest_INFO_from_iRODS.txt
#   ├── bams
#   │   └── WES_UNFILT
#   ├── fastqs
# 
#Author: Martin Del Castillo Velasco-Herrera - mdc1@sanger.ac.uk
###################################################################
progname<-"PDX_bwa_mem_mapping_jobs_from_master_manif.R"

#Loading Libraries
suppressMessages(library(optparse))

#Get the options for tthe script
####### Set the options list of the program
option_list<- list(
  make_option(c("--manifest"), action="store_true", default=NA, type="character", help="This is the project iRODs manifest file name generated by sequencescape project ID number using Build_manifest_from_irods_cram_information.R
                IT should be locatet in the folder /projectdir/manifests/ for this command to work"), 
  make_option(c("--projectdir" ), action="store_true",type="character", default=NA, help="This is the location where the main project folder is (if provided, else it will assume the current directory)"),
  make_option(c("--referencedir" ), action="store_true",type="character", default=NA, help="This is the location where the mouse reference is located")
)
#### Indicate that the file follows to the options used
parser<- OptionParser(usage = paste(progname, " [options] --manifest SeqscapeID_canappsID_master_manifest_wmetadata_for_proj.txt --projectdir  /My_project_folder/ --referencedir /PATH/To/my/reference ",sep=""), option_list=option_list)
arguments<- parse_args(parser, positional_arguments = 0) # no files after the options shoudl be provided

# get command line options, if help option encountered print help and exit,
# otherwise if options not found on command line then set defaults,
opt<-arguments$options

#Verify that the projectdir was set
if(is.na(opt$projectdir)){
  projectdir<-paste(getwd(),"/",sep="")
} else {
  projectdir<-opt$projectdir
}

#Verify that the referencedir was set
if(is.na(opt$referencedir)){
  stop("No reference directory was provided !")
} else {
  refdir<- opt$referencedir
  if(!dir.exists(refdir) ){
    stop(paste0("The reference directory provided doesn't exist ! : ", refdir ))
  }
}

#Verify that the input manifest exitst was set 
manifdir<-file.path(projectdir, "metadata", "manifests") 
scriptsdir<-file.path(projectdir, "scripts", "pdx_processing" )
manif_name<-file.path(manifdir,opt$manifest)
if(!file.exists(manif_name)){
  stop(paste("The manifest file:", manif_name, "doesn't exits. Check file or see the help for requirements", sep=""))
}
#Getting the study id
studyid<-unlist(strsplit(opt$manifest, split = "_",fixed = T) )[1]

#Checking if samtools is visible
if (length(system("which samtools &> /dev/null", intern=T))==0 ) {
  warning("samtools is not available.")
} else { # Print samtools version
  write(paste( "Found samtools version:"), stdout())
  write(paste( system("samtools --version | head -n 2", intern=T), sep=" "), stdout())
}


#Write some output starting message
write(paste("#########################################\n", progname, date(),"The study ID found from manifest name is: ",studyid, "\nThe Manifest provided:", opt$manifest,"\n",
            "The projectdir provided is:", projectdir, "\n", "The project manifest found is: ",manif_name, "\n",sep=" "), stdout())

#######################################################################
#Defining the directories for the analyses and GENERAL VARIABLES

#Set of folders for the outputs
fqfolder<-file.path(projectdir, "fastqs")
logfolder<-paste(projectdir, "/logs", sep="")
#create the folder for the mapping results 
bamdir<-file.path(projectdir,"bams", "WES_UNFILT")
bwamem_maplogdir<-file.path(logfolder, "bwamem_logs")
persampl_mergelogdir<-file.path(logfolder, "persamp_merge_logs")


#Make the folders
dir.create(bamdir, recursive = TRUE)
dir.create(bwamem_maplogdir, recursive = TRUE)
dir.create(persampl_mergelogdir, recursive = TRUE)

#PAHTS to the programs required
samtools<-"samtools"
#bwa mem module for the latest version of bwa mem module 
#bwamem<-"module load bwa-0.7.17 ; bwa mem"
bwamem<-"bwa mem"
# BWA MEM reference index directory and GTF for -NOD/ShiLtJ_V1 reference for PDX filtering 
bwamem_nodv1_ref_fasta<-file.path(refdir, "genome.fa")
if(!file.exists(bwamem_nodv1_ref_fasta)){
  stop( paste0("The reference file doesn't exist at the designated location: "))
}

#############################################################################################
# 1. Read the file  manifest file  ------------------
manif<-read.csv(manif_name, header=TRUE, stringsAsFactors=FALSE, sep="\t")
manif$SANGER_SAMPLE_ID<-manif$sample
manif$sample <- manif$sample_supplier_name
#Add the fastq names
manif$fastq1 <- file.path(fqfolder, manif$sample, paste(manif$sample, gsub(".cram","", manif$cram, fixed = T), "R1.fastq.gz", sep="_")) 
manif$fastq2 <- file.path(fqfolder, manif$sample, paste(manif$sample, gsub(".cram","", manif$cram, fixed = T), "R2.fastq.gz", sep="_")) 

#############################################################################################
# 2. Create the sh file that will submit  the jobs for mouse mapping from FASTA transformation  ------------------

#############################################################################################
# 3. Create the per sample manifest   ------------------
#In this case since the samples were sequenced into a single lane  there is no need get a manifest with all the information per unique sample

#############################################################################################
# 4. Create the sh file that will submit the mapping with STAR agains NOD/ShiLtJ_V1 for Tumours only------------------
#bsub -q long -J "PD53330a" -M 16000 -R"select[mem>16000] rusage[mem=16000] span[ptile=12]" -n 12  
#-o /lustre/scratch116/casm/team113/projects/4624_TSG_knockout_in_hiPSCs_Jyoti_Clara/STAR2_5_bams/logs/4624STDY6739364STAR_run.o 
#-e /lustre/scratch116/casm/team113/projects/4624_TSG_knockout_in_hiPSCs_Jyoti_Clara/STAR2_5_bams/logs/4624STDY6739364STAR_run.e 
#'/software/pkgg/bwa/0.7.17/bin/bwa mem 
#-t 12 
# -Y 
# -K 100000000
# -R "@RG\tID:COL_${INDEX}\tSM:COL_${INDEX}"
# "ID:\tLB:6735728\tSM:PD5330a\tPL:ILLUMINA"

#/lustre/scratch124/casm/team113/ref/NOD_ShiLtJ_V1_PDX_ref/genome.fa 
# PD53330a_R1.fastq.gz
# PD53330a_R2.fastq.gz
# bwa mem genome.fa PD53330a_41760_1#2_R1.fastq.gz PD53330a_41760_1#2_R2.fastq.gz | samtools sort -m 1G -@ 10 -o PD53330a_NOD_PDXV1.aln.sort.out.bam - ;
# samtools index PD53330a_NOD_PDXV1.aln.sort.out.bam

#Identify the tumour samples since this will be the only ones to process for further analyses
tumour_pos<- manif$Proc_as_PDX=="Y"
tumour_sample_names<- unique(manif$sample[tumour_pos])

#Create the two directories that will be created to store the mouse mapped files. 
refnam<-"NOD_PDXV1"
nodv1_bamfile_dir<- file.path(projectdir,"bams", "NOD_mapping", refnam)
dir.create(nodv1_bamfile_dir, recursive = T)

#Generate the set of mapping commands - NODv1
i<-1
bamfile<-NULL
cmds<-NULL
bam_path<-NULL
mergedbam_path<-NULL
tmanif<-manif
for (i in 1:dim(tmanif)[1]){
  sdoutf<-NULL
  serrf<-NULL
  temp<-tmanif[i,]
  cmd<-NULL
  bam_prefix<-NULL
  bam_sampdir<-NULL
  tbamp<-NULL
  tmerbamp<-NULL
  # give name to errorfiles
  sdoutf<-file.path(bwamem_maplogdir, paste("bwamem_mapping_log_",refnam,"_", i,".o", sep = ""))
  serrf<-file.path(bwamem_maplogdir, paste("bwamem_mapping_log_",refnam,"_", i,".e", sep = ""))
  bam_sampdir<-file.path(nodv1_bamfile_dir, temp$sample)
  print(paste("Creating the output directory for mapping of sample", temp$sample))
  dir.create(bam_sampdir)
  bam_prefix<-file.path(bam_sampdir, paste(temp$sample, "_", refnam, sep=""))
  tbamp<-paste(bam_prefix, paste(temp$id_run,"_",temp$lane, "#",temp$tag_index,sep=""), ".aln.sort.out.bam", sep="")
  tmerbamp<-paste(bam_prefix, ".sample.aln.sort.out.bam", sep="") # For the merged bamfile for the sample
  #Section that creates the farm command to submit the job the job submission 
  # should look like this
  #Memory
  cmd<-as.character(paste("bsub -q long -M 36000 -R",shQuote("select[mem>36000] rusage[mem=36000] span[hosts=1]") , " -n 12 -o ", sdoutf," -e ",serrf,
                          " '", bwamem, " -t 12 -Y -K 100000000 -R ",
                          #Read group 
                          "\"",  paste("@RG",paste(paste("ID:",temp$id_run,"_",temp$lane, "#",temp$tag_index,sep=""),paste("LB:",temp$library_id,sep=""), paste("SM:",temp$sample_supplier_name,sep=""), "PL:ILLUMINA", sep="\\t"), sep="\\t"), "\"",
                          #Reference genome 
                          " ", bwamem_nodv1_ref_fasta, " ",
                          temp$fastq1, " ", temp$fastq2 ,
                          " | ", samtools , " sort -m 1G -@ 10 -o ", tbamp, " - ; ",
                          samtools, " index ", tbamp,
                          " '",
                          sep=""))  
  #ADD the jobs
  cmds<-c(cmds,cmd)
  #add to the list of bampaths for bams per RUN_Lane#TagID
  bam_path<-c(bam_path, tbamp)
  #add to the list of bampaths for bams per RUN_Lane#TagID
  mergedbam_path<- c(mergedbam_path, tmerbamp)
}

#Add the bampaths to the manifes
manif$bam_run_lane_tid_path_nodv1<-bam_path
manif$bam_psample_path_nodv1<-mergedbam_path
#Create the sh file 
write.table(c("#!/bin/sh", cmds[tumour_pos]), file=file.path(scriptsdir, paste("bwamem_mapping_perlanrun_to_",refnam, "_tum_only_jobs.sh", sep="")), quote = F, col.names = F, row.names = F, sep = "\n")

#############################################################################################
# 5.Create the jobs for file merging      ------------------

# Create the two directories that will be created to store the mouse mapped files. 
cmds_nodv1<-NULL
for(i in 1:length(tumour_sample_names)){
  print(paste0("Processing sample ", tumour_sample_names[i], "NodV1" ))
  tempsamp<-NULL
  cmd<-NULL
  sdoutf<- NULL
  serrf<- NULL
  #Assing job output names 
  sdoutf<-file.path(persampl_mergelogdir, paste("samtools_psamp_merge_log_nodv1_", i,".o", sep = ""))
  serrf<-file.path(persampl_mergelogdir, paste("samtools_psamp_merge_log_nodv1_", i,".e", sep = ""))
  tempsamp<- tumour_sample_names[i]
  temp<-manif[ manif$sample==tempsamp,  ]
  #Command for Nodv1 per sample file and index
  cmd<-as.character(paste("bsub -q normal -M 16000 -R",shQuote("select[mem>16000] rusage[mem=16000] span[hosts=1]") , " -n 8 -o ", sdoutf," -e ",serrf,
                          " '", samtools,  " merge -@ 8 -o ", temp$bam_psample_path_nodv1[1], " ", 
                          #the names of the lane bam files NODV1
                          paste(temp$bam_run_lane_tid_path_nodv1, collapse = " "),  " ; ",
                          # index the bam file 
                          samtools, " index ", temp$bam_psample_path_nodv1[1],
                          " '",
                          sep=""))  
  #Append the command to the list of NODv1 merging
  cmds_nodv1<- c(cmds_nodv1, cmd)
  
}

#Generate the .sh files with the submissions 
write.table(c("#!/bin/sh", cmds_nodv1), file=file.path(scriptsdir, paste("samtools_psample_merge_nodv1_tum_only_jobs.sh", sep="")), quote = F, col.names = F, row.names = F, sep = "\n")


#############################################################################################
# 7. Write the manifest with the path files      ------------------

#Remove the per sample files for the non tumour samples 
#Add the bampaths to the manifest
manif$bam_run_lane_tid_path_nodv1[!tumour_pos]<-NA
manif$bam_psample_path_nodv1[!tumour_pos]<-NA
#Add location of the initial unfiltered bam file
manif$unfilt_psample_bam_path<-file.path(projectdir,"bams","WES_UNFILT", manif$sample,
 paste0(manif$sample, ".sample.dupmarked.bam" ))

#Name of the outfile manif
out_manif_name<-file.path(manifdir,gsub(pattern = ".txt", "_psamp_mouse.txt", opt$manifest))
write.table(manif, file=out_manif_name, quote = F, col.names = T, row.names = F, sep = "\t")

