#!/software/R-4.1.0/bin/Rscript 
#################################################################
# run_htseqcounts_from_RNAseq_master_manif.R
# This is s the 
#
#  The Project folder should have the following  structure before running this program:
#  /My_project_folder/
#   ├── manifests/My_cram_manifest_INFO_from_iRODS.txt
#   ├── BAMS
#   │   └── WES_UNFILT
#   ├── fastqs
# The program will create the  following directories:
# /My_project_folder/
#     /fastqs
#       /PDID
#     /logs
#       /cramtofastqs
#       /STAR_logs
#       /htseq_logs
#       /RSeQC_logs
# 
# 
# 
#Created: 08/2022 
#Author: Martin Del Castillo Velasco-Herrera - mdc1@sanger.ac.uk
###################################################################
progname<-"run_htseqcounts_from_RNAseq_master_manif.R"

#Loading Libraries
suppressMessages(library(optparse))

#Get the options for tthe script
####### Set the options list of the program
option_list<- list(
  make_option(c("--manifest"), action="store_true", default=NA, type="character", help="This is the project iRODs manifest file name generated by sequencescape project ID number using Build_manifest_from_irods_cram_information.R
                IT should be locatet in the folder /projectdir/manifests/ for this command to work"), 
  make_option(c("--projectdir" ), action="store_true",type="character", default=NA, help="This is the location where the main project folder is (if provided, else it will assume the current directory)"), 
  make_option(c("--outdir" ), action="store_true",type="character", default=NA, help="This is the location where output folder with per sample folders and filtered BAMS will be located ")
)
#### Indicate that the file follows to the options used
parser<- OptionParser(usage = paste(progname, " [options] --manifest SeqscapeID_canappsID_master_manifest_wmetadata_for_proj_psamp_mouse.txt --projectdir  /My_project_folder/ --outdir ",sep=""), option_list=option_list)

arguments<- parse_args(parser, positional_arguments = 0) # no files after the options shoudl be provided

#This is for testing purposes only
arguments<- parse_args(parser,
                         args=c("--manifest","6591_2775_master_manifest_wmetadata_for_proj_wNOD_bams_xfb.txt",
                                "--projectdir","/lustre/scratch124/casm/team113/projects/6591_PDX_models_Latin_America_RNAseq",
                                "--outdir","/lustre/scratch124/casm/team113/projects/6591_PDX_models_Latin_America_RNAseq/htseq_counts"
                               ), #For tests
##                        # args=c("--manifest","6591_2775_master_manifest_wmetadata_for_proj_wNOD_bams.txt","--projectdir","/Users/mdc1/Desktop/team113sc124/projects/6591_PDX_models_Latin_America_RNAseq"), #For tests from PC
                         positional_arguments = 0) # no files after the options should be provided

# get command line options, if help option encountered print help and exit,
# otherwise if options not found on command line then set defaults,
opt<-arguments$options


#Verify that the outidr was set
if(is.na(opt$projectdir)){
  projectdir<-paste(getwd(),"/",sep="")
} else {
  projectdir<-opt$projectdir
}
#Verify that the outidr was set and create it if it doesn't exists
if(is.na(opt$outdir)){
  outdir<-getwd()
} else {
  outdir<-opt$outdir
}
if(!dir.exists(outdir)){ # Create the outdir if it doesn't exist
  dir.create(outdir, recursive = T )
}

#Verify that the input manifest exitst was set 
manif_name<-file.path(projectdir,"manifests",opt$manifest)
if(!file.exists(manif_name)){
  stop(paste("The manifest file:", manif_name, "doesn't exits. Check file or see the help for requirements", sep=""))
}
#Getting the study id
studyid<-unlist(strsplit(opt$manifest, split = "_",fixed = T) )[1]

#Write some output starting message
write(paste("#########################################\n", progname, date(),"The study ID found from manifest name is: ",studyid, "\nThe Manifest provided:", opt$manifest,"\n",
            "The projectdir provided is:", projectdir, "\n", "The project manifest found is: ",manif_name, "\n", "The outputdir found is: ", outdir,sep=" "), stdout())

#######################################################################
#Defining the directories for the analyses and GENERAL VARIABLES

#Set of folders for the outputs
logfolder<-paste(projectdir, "/logs", sep="")
manifdir<-file.path(projectdir, "manifests")
#Create the folders for the htseq-counts and logs
countdir<-file.path(projectdir, "htseq_counts") 
htseq_logdir<-file.path(logfolder, "htseq_logs")

#create the folder for the counts
dir.create(countdir, recursive = TRUE)
dir.create(htseq_logdir, recursive = TRUE)


#Make the folders
dir.create(xfilt_logdir, recursive = TRUE)

#PAHTS to the programs required
#HTseq-counts version 0.13.5
htseq<-"/software/team113/users/mdc1/python/bin/htseq-count"
hum_ref_gtf<-"/lustre/scratch124/casm/team78pipelines/reference/human/GRCh38_ERCC92_RNA2021/cgpRna/103/ensembl.gtf"


#############################################################################################
# 1. Read the file  manifest file  ------------------
manif<-read.csv(manif_name, header=TRUE, stringsAsFactors=FALSE, sep="\t")
manif$sample <- manif$sample_supplier_name

# 1. Specify which BAM file was the final one used for the analysis per sample  ------------------
manif$final_bam_used_counts<- manif$xfilt_bam_psample_path_nodv1
manif$final_bam_used_counts[!grepl("PDX", manif$sample_type, fixed=TRUE)]<- manif$unfilt_psample_bam_path[!grepl("PDX", manif$sample_type, fixed=TRUE)]

#############################################################################################
# 5. Generate the sh script that is going to send the HTSeq-counts count jobs using the final_bam_used_counts bamfile location   ------------------
#Example job submission line
#bsub -q normal -o ./logs/htseq/htseq_count_test_%J.o -e ./logs/htseq/htseq_count_test_%J.e -M16000 -R"select[mem>16000] rusage[mem=16000] span[hosts=1]" -n 8 '/software/team113/users/mdc1/python/bin/htseq-count 
#Generate the set of counts command
i<-0
cmds<-NULL
htseqcounts_path<-NULL
for (i in 1:dim(manif)[1]){
  sdoutf<-NULL
  serrf<-NULL
  temp<-manif[i,]
  cmd<-NULL
  tcountsfile<-NULL
  # give name to error and stdout files
  sdoutf<-file.path(htseq_logdir, paste("HTSeq_counts_log_",i,".o", sep = ""))
  serrf<-file.path(htseq_logdir, paste("HTSeq_counts_log_",i,".e", sep = ""))
  tcountsfile<-file.path(countdir,paste(temp$sample, "_htseq_counts_srverse.gz", sep=""))
  #Section that creates the farm command to submit the job the job submission 
  # should look like this
  #
  cmd<-as.character(paste("bsub -q normal -M 18000 -R",shQuote("select[mem>18000] rusage[mem=18000] span[hosts=1]") , " -n 12 -o ", sdoutf," -e ",serrf,
                          " ' module load bgzip/1.13 ; ", htseq, " --format bam --order pos -m union --stranded reverse --idattr gene_id --type exon ",
                          temp$final_bam_used_counts," ",hum_ref_gtf, " | gzip -c >",tcountsfile, 
                          "'",sep=""))  
  #ADD the jobs
  cmds<-c(cmds,cmd)
  #add to the list of bampaths
  htseqcounts_path<-c(htseqcounts_path, tcountsfile)
}

#Add the bampaths to the manifes
manif$htseqcounts_path<-htseqcounts_path
#Create the sh file 
write.table(c("#!/bin/sh", cmds), file=file.path(projectdir,"scripts", "htseqcount_countsing_jobs.sh"), quote = F, col.names = F, row.names = F, sep = "\n")


#############################################################################################
# 3. update manifest 
#Name of the outfile manif
out_manif_name<-file.path(projectdir,"manifests",gsub(pattern = ".txt", "_wcount.txt", opt$manifest))
write.table(manif, file=out_manif_name, quote = F, col.names = T, row.names = F, sep = "\t")

