#!/usr/bin/env Rscript
#################################################################
# run_Xenofilter_from_WES_master_manif.R
# This script takes an MAster manifest generated from 6591_master_manifest_creation_RNAseq.R to create series of files with a series of command job submission to import, sort, transform to fastq, map with STAR
# and generate counts with htseqcount plus some quality controls for the read nucleotide composition and TIN (currently not working)
# All of this with the Rat genome assembly (Rnor6.0 plus 92 ERCC spike ins) 
#
#  The Project folder should have the following  structure before running this program:
#  /My_project_folder/
#   ├── manifests/My_cram_manifest_INFO_from_iRODS.txt
#   ├── BAMS
#   │   └── WES_UNFILT
#   ├── fastqs
# The program will create the  following directories:
# /My_project_folder/
#     /fastqs
#       /PDID
#     /logs
#       /cramtofastqs
#       /STAR_logs
#       /htseq_logs
#       /RSeQC_logs
# 
# 
#Author: Martin Del Castillo Velasco-Herrera - mdc1@sanger.ac.uk
###################################################################
progname<-"run_Xenofilter_from_WES_master_manif.R"

#Loading Libraries
suppressMessages(library(optparse))

#Get the options for tthe script
####### Set the options list of the program
option_list<- list(
  make_option(c("--manifest"), action="store_true", default=NA, type="character", help="This is the project iRODs manifest file name generated by sequencescape project ID number using Build_manifest_from_irods_cram_information.R
                IT should be locatet in the folder /projectdir/manifests/ for this command to work"), 
  make_option(c("--projectdir" ), action="store_true",type="character", default=NA, help="This is the location where the main project folder is (if provided, else it will assume the current directory)"), 
  make_option(c("--outdir" ), action="store_true",type="character", default=NA, help="This is the location where output folder with per sample folders and filtered BAMS will be located ")
)
#### Indicate that the file follows to the options used
parser<- OptionParser(usage = paste(progname, " [options] --manifest SeqscapeID_canappsID_master_manifest_wmetadata_for_proj_psamp_mouse.txt --projectdir  /My_project_folder/ --outdir ",sep=""), option_list=option_list)

arguments<- parse_args(parser, positional_arguments = 0) # no files after the options shoudl be provided

#This is for testing purposes only
# arguments<- parse_args(parser,
#                        args=c("--manifest","6633_2729_master_manifest_wmetadata_for_proj_psamp_mouse.txt",
#                               "--projectdir","/lustre/scratch124/casm/team113/projects/6633_PDX_models_Latin_America_WES",
#                               "--outdir","/lustre/scratch124/casm/team113/projects/6633_PDX_models_Latin_America_WES/BAMS/WES_xfilt"
#                              ), #For tests
#                        # args=c("--manifest","6591_2775_master_manifest_wmetadata_for_proj.txt","--projectdir","/Users/mdc1/Desktop/team113sc124/projects/6591_PDX_models_Latin_America_RNAseq"), #For tests from PC
#                        positional_arguments = 0) # no files after the options should be provided

# get command line options, if help option encountered print help and exit,
# otherwise if options not found on command line then set defaults,
opt<-arguments$options

#Verify that the outidr was set
if(is.na(opt$projectdir)){
  projectdir<-paste(getwd(),"/",sep="")
} else {
  projectdir<-opt$projectdir
}
#Verify that the outidr was set and create it if it doesn't exists
if(is.na(opt$outdir)){
  outdir<-getwd()
} else {
  outdir<-opt$outdir
}
if(!dir.exists(outdir)){ # Create the outdir if it doesn't exist
  dir.create(outdir, recursive = T )
}

#Verify that the input manifest exitst was set 
manifdir<-file.path(projectdir, "metadata", "manifests")
manif_name<-file.path(manifdir,opt$manifest)
if(!file.exists(manif_name)){
  stop(paste("The manifest file:", manif_name, "doesn't exits. Check file or see the help for requirements", sep=""))
}
#Getting the study id
studyid<-unlist(strsplit(opt$manifest, split = "_",fixed = T) )[1]

#Write some output starting message
write(paste("#########################################\n", progname, date(),"The study ID found from manifest name is: ",studyid, "\nThe Manifest provided:", opt$manifest,"\n",
            "The projectdir provided is:", projectdir, "\n", "The project manifest found is: ",manif_name, "\n", "The outputdir found is: ", outdir,sep=" "), stdout())

#######################################################################
#Defining the directories for the analyses and GENERAL VARIABLES

#Set of folders for the outputs
logfolder<-paste(projectdir, "/logs", sep="")
scriptsdir<-file.path(projectdir, "scripts", "pdx_processing")
xfilt_logdir<-file.path(logfolder, "xfilt_logdir")

#Make the folders
dir.create(xfilt_logdir, recursive = TRUE)

#PAHTS to the programs required
#run_xenofilter<-"/lustre/scratch124/casm/team113/projects/6633_PDX_models_Latin_America_WES/scripts/bam_Xenofilter.R"
run_xenofilter<-file.path(scriptsdir, "bam_Xenofilter.R")


#############################################################################################
# 1. Read the file  manifest file  ------------------
manif<-read.csv(manif_name, header=TRUE, stringsAsFactors=FALSE, sep="\t")

#############################################################################################
# 2. Create the sh file that will submit the filter of the BAM files with the NOD/ShiLtJ_V1 for Tumours only------------------
# bsub -q normal -M 64000 -R"select[mem>64000] rusage[mem=64000] span[hosts=1]" -n 2 '

#Identify the tumour samples since this will be the only ones to process for further analyses
tumour_pos<- manif$Proc_as_PDX=="Y"
tumour_sample_names<- unique(manif$sample[tumour_pos])

#Create the two directories that will be created to store the mouse mapped files. 
refname_nod1<-"NOD_PDXV1"

#Create the expected name for the ouputs 
dir.create(file.path(outdir,refname_nod1), recursive = T)

#Then generate the name of the ouput filtered file 
manif$xfilt_bam_psample_path_nodv1<- file.path(outdir,refname_nod1, manif$sample,"Filtered_bams", paste0(manif$sample, ".sample.dupmarked.mXfilt_Filtered.bam")) 

# Create the two directories that will be created to store the mouse mapped files. 
cmds_nodv1<-NULL
for(i in 1:length(tumour_sample_names)){
  print(paste0("Processing sample ", tumour_sample_names[i], " NodV1" ))
  tempsamp<-NULL
  cmd<-NULL
  sdoutf<- NULL
  serrf<- NULL
  thbam<-NULL #Temp human bam
  tmbam<-NULL #Temp mouse bam
  #Assing job output names 
  sdoutf<-file.path(xfilt_logdir, paste("xenofilter_log_nodv1_", i,".o", sep = ""))
  serrf<-file.path(xfilt_logdir, paste("xenofilter_log_nodv1_", i,".e", sep = ""))
  #Sample analysing
  tempsamp<- tumour_sample_names[i]
  temp<-manif[ manif$sample==tempsamp,  ]
  thbam<- temp$unfilt_psample_bam_path[1]
  tmbam<- temp$bam_psample_path_nodv1[1]
  #Command for Nodv3 per sample file and index Min 64GBs RAM
  cmd<-as.character(paste("bsub -q normal -M 120000 -R",shQuote("select[mem>120000] rusage[mem=120000] span[hosts=1]") , " -n 2 -o ", sdoutf," -e ",serrf,
                          " 'Rscript ", run_xenofilter, " --sample_name ", tempsamp,
                              " --human_bam ", thbam,
                              " --mouse_bam ", tmbam,
                              " --outdir ", file.path(outdir,refname_nod1),
                              " --ncpu ", "1",
                          " '",
                          sep=""))  
  #Append the command to the list of NODv1 merging
  cmds_nodv1<- c(cmds_nodv1, cmd)
}

#Generate the .sh files with the submissions 
write.table(c("#!/bin/sh", cmds_nodv1), file=file.path(scriptsdir, paste("xenofilter_nodv1_tum_only_jobs.sh", sep="")), quote = F, col.names = F, row.names = F, sep = "\n")

#############################################################################################
# 3. update manifest 
#Name of the outfile manif
out_manif_name<-file.path(manifdir,gsub(pattern = ".txt", "_xfb.txt", opt$manifest))
write.table(manif, file=out_manif_name, quote = F, col.names = T, row.names = F, sep = "\t")

